lib/wash/exe/eval.js

```diff
@@ -0,0 +1,195 @@
+
+/** @typedef JsExecuteContext$$module$axiom$fs$js$execute_context */
+var JsExecuteContext;
+
+/** @typedef StatResult$$module$axiom$fs$stat_result */
+var StatResult;
+
+/**
+ * @constructor
+ * A partial clone of GNU jsEval.
+ *
+ * @param {JsExecuteContext} executeContext
+ */
+var Eval = function(executeContext) {
+  this.executeContext = executeContext;
+
+  /**
+   * @private @type {Termcap}
+   */
+  this.tc_ = new Termcap();
+
+  this.exit = false;
+
+  this.promptString_ = executeContext.getArg('prompt-string', '');
+  this.promptVars_ = null;
+
+  this.inputHistory = [];
+};
+
+export {Eval};
+
+/**
+ * @param {JsExecuteContext} cx
+ * @return {void}
+ */
+export var main = function(cx) {
+
+  cx.ready();
+
+  var jsEval = new Eval(cx);
+
+  // TODO(grv): implement input-history for eval command.
+  jsEval.inputHistory = cx.getArg('input-history', []);
+
+  jsEval.setPrompt();
+
+  jsEval.readEvalPrintLoop().then(function(value) {
+    cx.closeOk(value);
+  }).catch(function(err) {
+    cx.closeError(err);
+  });
+};
+
+/**
+ * Read-eval-print-loop.
+ *
+ * @return {Promise<*>} Resolves to the value of the final evaluation.
+ */
+Eval.prototype.readEvalPrintLoop = function() {
+  return this.readEvalPrint('').then(
+    function(value) {
+      if (this.exit) {
+        return Promise.resolve();
+      }
+      if (this.executeContext.isEphemeral('Ready'))
+        return this.readEvalPrintLoop();
+
+      return Promise.resolve(value);
+    }.bind(this)
+  ).catch(
+    function(value) {
+      return Promise.reject(value);
+    }.bind(this)
+  );
+};
+
+/**
+ * Read a single line of input, eval it, print the result or an error.
+ *
+ * @return {Promise<*>} Resolves to result of the evaluation.
+ */
+Eval.prototype.readEvalPrint = function(inputString) {
+  var input = inputString;
+  var promptString = this.promptString_;
+
+  if (inputString !== '') {
+    promptString = '';
+  }
+
+  return this.read(promptString).then(
+    function(result) {
+      if (result === null || result === 'exit') {
+          this.exit = true;
+        return Promise.resolve();
+      }
+
+      if (typeof result != 'string') {
+        return Promise.reject(new AxiomError.Runtime(
+            'Unexpected type from readline: ' + (typeof result)));
+      }
+
+      input += result;
+      return this.evaluate(input).then(
+        function(value) {
+          if (typeof value !== 'undefined' && value !== null) {
+            this.executeContext.stdout.write(value + '\n');
+          }
+          return Promise.resolve(value);
+        }.bind(this));
+    }.bind(this)
+  ).catch(
+    function(error) {
+      // We currently do not support multi line input in readline. This is a
+      // proxy to find if the error is a syntax error in code or the user has
+      // not finished typing.
+      // TODO(grv): Do a better detection of syntax error vs incomplete code.
+      if (error && error.message == 'Unexpected end of input') {
```

Comment:
This seems fragile. Where is this message coming from? Can't we wrap it at the call site and throw a custom exception?

At the very least, I would add a note explaining where the message is coming from.